In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
df = pd.read_csv('creditcard.csv')

In [3]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [7]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

distribution of normal and fraud transactions

In [8]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [9]:
#sepeate data 
normal = df[df.Class == 0]
fraud = df[df.Class == 1]

In [10]:
normal.shape

(284315, 31)

In [11]:
fraud.shape

(492, 31)

building new normal dataset containing similar distributions of normal and fraud transactions

In [12]:
normal_sample = normal.sample(n=492)

In [13]:
#concatinating the two dataset
new_dataset = pd.concat([normal_sample,fraud],axis = 0)

In [14]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
262767  160637.0  0.612594  0.152592 -0.309627 -0.119733  0.806966 -0.440160   
101346   67803.0  1.301311  0.468738 -0.015289  0.907748  0.292125 -0.452431   
166889  118367.0  1.837468 -0.761769 -0.835587  0.095840 -0.374113  0.238155   
253809  156430.0 -0.120400  0.845400  0.297701 -0.692006  0.331591 -0.846500   
227602  145139.0 -2.075661  0.111866 -0.144546  0.079908  1.626611 -1.056263   

              V7        V8        V9  ...       V21       V22       V23  \
262767 -1.575931 -1.934770 -1.307684  ...  1.500575 -0.676076 -1.835391   
101346  0.326046 -0.284249 -0.232755  ... -0.005787  0.120158 -0.226296   
166889 -0.658886  0.222902  1.360517  ...  0.271242  0.648005  0.025380   
253809  0.890229 -0.075607 -0.161557  ... -0.191320 -0.487426  0.082339   
227602  1.421973  0.196238 -1.538231  ...  0.396345  0.426491 -0.447482   

             V24       V25       V26       V27       V28  Amount  Class  
262767 -0.848020  0.366664  0.276758  0.351293  0.289686   25.00      0  
101346 -0.406401  0.859488 -0.282011  0.024206  0.015890    6.00      0  
166889  0.158309 -0.084475 -0.586499  0.024423 -0.029686   89.00      0  
253809  0.054926 -0.562138  0.103910  0.104814  0.084377    7.99      0  
227602 -0.310733  1.576384 -0.361898 -0.173618 -0.209652  138.00      0  

[5 rows x 31 columns]

In [15]:
new_dataset.shape

(984, 31)

spliting the new_dataset into x and y

In [16]:
x = new_dataset.drop(columns = 'Class',axis = 1)
y = new_dataset['Class']

In [17]:
x

Time        V1        V2        V3        V4        V5        V6  \
262767  160637.0  0.612594  0.152592 -0.309627 -0.119733  0.806966 -0.440160   
101346   67803.0  1.301311  0.468738 -0.015289  0.907748  0.292125 -0.452431   
166889  118367.0  1.837468 -0.761769 -0.835587  0.095840 -0.374113  0.238155   
253809  156430.0 -0.120400  0.845400  0.297701 -0.692006  0.331591 -0.846500   
227602  145139.0 -2.075661  0.111866 -0.144546  0.079908  1.626611 -1.056263   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V20       V21       V22  \
262767 -1.575931 -1.934770 -1.307684  ...  0.082796  1.500575 -0.676076   
101346  0.326046 -0.284249 -0.232755  ...  0.001406 -0.005787  0.120158   
166889 -0.658886  0.222902  1.360517  ... -0.098800  0.271242  0.648005   
253809  0.890229 -0.075607 -0.161557  ... -0.080029 -0.191320 -0.487426   
227602  1.421973  0.196238 -1.538231  ...  0.332053  0.396345  0.426491   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  1.252967  0.778584 -0.319189   
280143 -1.413170  0.248525 -1.127396  ...  0.226138  0.370612  0.028234   
280149 -2.234739  1.210158 -0.652250  ...  0.247968  0.751826  0.834108   
281144 -2.208002  1.058733 -1.632333  ...  0.306271  0.583276 -0.269209   
281674  0.223050 -0.068384  0.577829  ... -0.017652 -0.164350 -0.295135   

             V23       V24       V25       V26       V27       V28  Amount  
262767 -1.835391 -0.848020  0.366664  0.276758  0.351293  0.289686   25.00  
101346 -0.226296 -0.406401  0.859488 -0.282011  0.024206  0.015890    6.00  
166889  0.025380  0.158309 -0.084475 -0.586499  0.024423 -0.029686   89.00  
253809  0.082339  0.054926 -0.562138  0.103910  0.104814  0.084377    7.99  
227602 -0.447482 -0.310733  1.576384 -0.361898 -0.173618 -0.209652  138.00  
...          ...       ...       ...       ...       ...       ...     ...  
279863  0.639419 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00  
280143 -0.145640 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76  
280149  0.190944  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89  
281144 -0.456108 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00  
281674 -0.072173 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53  

[984 rows x 30 columns]

In [18]:
y

262767    0
101346    0
166889    0
253809    0
227602    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64

splitting dataset into training and testing

In [19]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size= 0.2, stratify=y,random_state=42)

In [20]:
print(x_train,x_test,y_train,y_test)

            Time         V1         V2         V3        V4         V5  \
221009  142392.0  -2.121936  -1.647227   1.374398  0.449036  -1.229553   
112590   72716.0  -1.322024   1.512386   0.625507  1.240638  -0.255313   
56703    47545.0   1.176716   0.557091  -0.490800  0.756424   0.249192   
150684   93888.0 -10.040631   6.139183 -12.972972  7.740555  -8.684705   
6882      8808.0  -4.617217   1.695694  -3.114372  4.328199  -1.873257   
...          ...        ...        ...        ...       ...        ...   
39183    39729.0  -0.964567  -1.643541  -0.187727  1.158253  -2.458336   
253362  156215.0  -1.467851  -0.984280   0.888707 -0.700243  -1.160602   
143335   85285.0  -6.713407   3.921104  -9.746678  5.148263  -5.151563   
85382    60750.0   1.268781  -0.961254   0.465823 -2.140585  -1.552236   
17317    28625.0 -27.848181  15.598193 -28.923756  6.418442 -20.346228   

              V6         V7         V8        V9  ...       V20       V21  \
221009  1.910142   2.648675   0.06

model training and testing

In [21]:
model = LogisticRegression()

In [22]:
model.fit(x_train,y_train)

LogisticRegression()

accuracy score

In [23]:
y_predict = model.predict(x_test)

In [25]:
acc = accuracy_score(y_test,y_predict)

In [26]:
acc

0.9238578680203046

so we got accuracy of 92% by training a model using logistic regression
